txpool/txpool.go

```diff
@@ -823,466 +531,508 @@ func (t *TxPool) validateTx(
-		heap:  newMinTxPriceHeapImpl(),
-	}
+// Returns the next expected nonce for this account.
+func (a *account) getNonce() uint64 {
+	return atomic.LoadUint64(&a.nextNonce)
 }
 
-// Required method definitions for the standard golang heap package
-type txPriceHeapImplBase struct {
-	txs []*pricedTx
+// Sets the next expected nonce for this account.
+func (a *account) setNonce(nonce uint64) {
+	atomic.StoreUint64(&a.nextNonce, nonce)
 }
 
-func (t txPriceHeapImplBase) Len() int { return len(t.txs) }
+// Pushes the transaction onto the enqueued queue.
+func (a *account) enqueue(tx *types.Transaction, demoted bool) error {
+	a.enqueued.lock(true)
+	defer a.enqueued.unlock()
 
-func (t txPriceHeapImplBase) Swap(i, j int) {
-	t.txs[i], t.txs[j] = t.txs[j], t.txs[i]
-	t.txs[i].index = i
-	t.txs[j].index = j
-}
+	// only accept low nonce if
+	// tx was demoted
+	if tx.Nonce < a.getNonce() &&
+		!demoted {
+		return ErrNonceTooLow
+	}
 
-func (t *txPriceHeapImplBase) Push(x interface{}) {
-	n := len(t.txs)
-	job := x.(*pricedTx)
-	job.index = n
-	t.txs = append(t.txs, job)
-}
+	// enqueue tx
+	a.enqueued.push(tx)
 
-func (t *txPriceHeapImplBase) Pop() interface{} {
-	old := *t
-	n := len(old.txs)
-	job := old.txs[n-1]
-	job.index = -1
-	t.txs = old.txs[0 : n-1]
-	return job
+	return nil
 }
 
-func (t txPriceHeapImplBase) Less(i, j int) bool {
-	return i < j
-}
+// Promote moves eligible transactions from enqueued to promoted.
+//
+// Eligible transactions are all sequential in order of nonce
+// and the first one has to have nonce less (or equal) to the account's
+// nextNonce.
+func (a *account) promote() uint64 {
+	a.promoted.lock(true)
+	a.enqueued.lock(true)
+	defer func() {
+		a.promoted.unlock()
+		a.enqueued.unlock()
+	}()
+
+	currentNonce := a.getNonce()
+	if a.enqueued.length() == 0 ||
+		a.enqueued.peek().Nonce > currentNonce {
+		// nothing to promote
+		return 0
+	}
+
+	promoted := uint64(0)
+	nextNonce := a.enqueued.peek().Nonce
+	for {
+		tx := a.enqueued.peek()
+		if tx == nil ||
+			tx.Nonce != nextNonce {
+			break
+		}
 
-// max price ordered tx heap implementation
-type maxTxPriceHeapImpl struct {
-	txPriceHeapImplBase
-}
+		// pop from enqueued
+		tx = a.enqueued.pop()
+
+		// push to promoted
+		a.promoted.push(tx)
 
-func newMaxTxPriceHeapImpl() heap.Interface {
-	return &maxTxPriceHeapImpl{
-		txPriceHeapImplBase: txPriceHeapImplBase{
-			make([]*pricedTx, 0),
-		},
+		// update counters
+		nextNonce += 1
+		promoted += 1
 	}
-}
 
-func (t maxTxPriceHeapImpl) Less(i, j int) bool {
-	if t.txs[i].from == t.txs[j].from {
-		return t.txs[i].tx.Nonce < t.txs[j].tx.Nonce
+	// only update the nonce map if the new nonce
+	// is higher than the one previously stored.
+	if nextNonce > currentNonce {
+		a.setNonce(nextNonce)
 	}
 
-	return t.txs[i].price.Cmp(t.txs[j].price) >= 0
+	return promoted
 }
 
-type minTxPriceHeapImpl struct {
-	txPriceHeapImplBase
+// Lookup map used to find transactions present in the pool
```

Comment:
You can move the lookupMap struct and all associated methods to a separate lookup_map.go file

Reply:
[Exported](https://github.com/0xPolygon/polygon-sdk/pull/287/commits/8f8c2a793b02c4744e33f3598b871ea3b1aa73e9)